src/libUtils/DnsUtils.cpp

```diff
@@ -0,0 +1,348 @@
+    return;
+  }
+
+  auto &ipList = dataListCache.ipList;
+
+  if (!QueryIpStrListFromDns(ipList, listType)) {
+    LOG_GENERAL(WARNING, "Failed to obtain ip list from "
+                             << dnsAddresses[listType]
+                             << ", try again on another DS epoch");
+    return;
+  }
+
+  vector<uint128_t> currentIpKeys;
+  currentIpKeys.reserve(ipList.size());
+
+  auto &pubKeyList = dataListCache.pubKeyList;
+  auto dnsAddress = dnsAddresses[listType];
+
+  // Adding new pubkeys to our dns cache
+  for (const auto &ipStr : ipList) {
+    auto ipKey = ConvertIpStringToUint128(ipStr);
+    currentIpKeys.emplace_back(ipKey);
+
+    if (pubKeyList.find(ipKey) != pubKeyList.end()) {
+      // Already exists, don't need to query again, unlikely to change pubkey
+      // for an ip
+      continue;
+    }
+
+    bytes pubKeyResult;
+    if (QueryPubKeyFromUrl(pubKeyResult, ipStr, listType)) {
+      pubKeyList[ipKey] = pubKeyResult;
+    }
+  }
+
+  // Remove pubkeys that are no longer in the dns list
+  for (auto itr = pubKeyList.begin(); itr != pubKeyList.end();) {
+    if (find(currentIpKeys.begin(), currentIpKeys.end(), itr->first) ==
+        currentIpKeys.end()) {
+      itr = pubKeyList.erase(itr);
+    } else {
+      ++itr;
+    }
+  }
+
+  dataListCache.isQuerying = false;
+  dataListCache.queryWaitCv.notify_all();
+}
+}  // namespace
+
+void InitDnsCacheList() {
+  dnsCacheListDataMap[DnsListType::UPPER_SEED];
+  dnsCacheListDataMap[DnsListType::L2L_DATA_PROVIDERS];
+  dnsCacheListDataMap[DnsListType::MULTIPLIERS];
+
+  dnsAddresses[DnsListType::UPPER_SEED] = UPPER_SEED_DNS;
+  dnsAddresses[DnsListType::L2L_DATA_PROVIDERS] = L2L_DATA_PROVIDERS_DNS;
+  dnsAddresses[DnsListType::MULTIPLIERS] = MULTIPLIER_DNS;
+
+  isShuttingDown = false;
+}
+
+void ShutDownDnsCacheList() { isShuttingDown = true; }
+
+void AttemptPopulateLookupsDnsCache() {
+  LOG_MARKER();
+  QueryDnsList(DnsListType::UPPER_SEED);
+  QueryDnsList(DnsListType::L2L_DATA_PROVIDERS);
+  QueryDnsList(DnsListType::MULTIPLIERS);
+}
+
+void AttemptPopulateLookupsDnsCacheImmediately(DnsListType listType) {
+  LOG_MARKER();
+
+  auto &dataListCache = dnsCacheListDataMap[listType];
+  if (!dataListCache.isQuerying) {
+    DetachedFunction(1, QueryDnsList, listType);
+    // else, wait for another thread to finish
+  }
+
+  // Timeout for result to go into cache
+  unique_lock<mutex> lock(dataListCache.queryWaitMutex);
+  if (dataListCache.queryWaitCv.wait_for(
+          lock, std::chrono::milliseconds(QUERY_DNS_TIMEOUT_MILLISECONDS)) ==
+      std::cv_status::timeout) {
+    LOG_GENERAL(WARNING, "Time out while querying " << dnsAddresses[listType]);
+  }
+}
+
+string GetPubKeyUrl(const std::string &ip, const std::string &url) {
+  auto tmpIp = ip;
+  std::replace(tmpIp.begin(), tmpIp.end(), '.', '-');
+  return "pub-" + tmpIp + url.substr(url.find_first_of('.'));
+}
+
+uint128_t ConvertIpStringToUint128(const string &ipStr) {
```

Comment:
We already have a utility function for this.
`https://github.com/Zilliqa/Zilliqa/blob/1571634cc97baf1d0b81310d1c72395be75a9ac4/src/libUtils/IPConverter.cpp#L88`

Reply:
got it, will use this